### PR TITLE
Update translator to 2021-03 lambdapi syntax

### PR DESCRIPTION
**Pull request by yannl35133**
I am new to both the project and scala, so the modifications may be messy or dodgy, but they should work.
I am particularly unsure of how to properly handle the test script.

### CODE REVIEW

**Comment by 01mf02** [comment]
Hello Yann,
thanks for your pull request! Can you please update the instructions in the README (in particular, the version of Lambdapi to be used) so I can test your changes?

**Comment by 01mf02** [comment]
What is your problem with the test script?

File: lib/Tools/dedukti_test [hunk 1]
```diff
@@ -4,8 +4,17 @@
...
+OUTPUT="${1:-isabelle_dedukti_test}"
 
-isabelle dedukti_import -O "$OUTPUT" Dedukti_Import
+if [ -d $OUTPUT ]; then
+    if [ ! -f "$OUTPUT/lambdapi.pkg" ]; then
+        echo "$OUTPUT is not a lambdapi package directory"
+        exit 1
+    fi
+else
+    $ISABELLE_LAMBDAPI init "${OUTPUT//"/"/.}"
```
Comment (01mf02) [inline lib/Tools/dedukti_test:15]: What does this line with the `//` do? Can you add a comment?

File: src/writers.scala [hunk 2]
```diff
@@ -41,9 +41,13 @@ trait IdentWriter
...
     name.forall(c => Symbol.is_ascii_letter(c) || Symbol.is_ascii_digit(c) || c == '_')
 
+  val suffix = "_"
+
   def escape(name: String): String =
     if (name.containsSlice("|}")) Exn.error("Bad name: " + Library.quote(name))
-    else "{|" + name + "|}"
+    else if ({val lastChar = name.charAt(name.length - 1); lastChar.isDigit || lastChar == '_'})
+        "{|" + name + suffix + "|}"
+      else "{|" + name + "|}"
```
Comment (01mf02) [inline src/writers.scala]: Why is this needed?

File: src/writers.scala [hunk 3]
```diff
@@ -177,48 +207,51 @@ class LPWriter(writer: Writer) extends LambdaPiWriter(writer)
...
+        write("symbol ");
         name(id)
         for (a <- args) { space; block { arg(a) } }
         for (ty <- ty) { colon; term(ty) }
-        dfn
+        colon_equal
         term(tm)
       case Syntax.Theorem(id, args, ty, prf) =>
-        write("theorem ");
+        write("symbol ");
```
Comment (01mf02) [inline src/writers.scala]: Are you sure that `symbol` is the right choice here?
Won't this rewrite the theorem name with its proof when used elsewhere?
(This slows down the checker!)

**Comment by 01mf02** [comment]
Just out of curiosity, did you test the export on some Isabelle theories, and if so, on which?

File: src/writers.scala (hunk 3 above)
Comment (fblanqui) [reply src/writers.scala]: Right, "theorem" should be replaced by "opaque symbol".

**Comment by fblanqui** [comment]
@01mf02 This is work in progress. Please, be patient. This PR is not going to be merged right now.

File: lib/Tools/dedukti_test (hunk 1 above)
Comment (yannl35133) [reply lib/Tools/dedukti_test:15]: My goal, and what `${OUTPUT//"/"/.}` should do, is replace every "/" in the argument by ".".
lambdapi expects a package root, which features directory nesting with "." instead of the Unix "/"

File: src/writers.scala (hunk 2 above)
Comment (yannl35133) [reply src/writers.scala]: lambdapi now rejects idents ending with numbers if they have leading zeros, and at some locations idents that must be escaped.
I have not yet well understood how this latter restriction works (the symbols may have escaped-only idents for example) so this is a cherry-pick of what I have found to cause problems : idents of shape `:ddd` (d: digit) used in lambda abstractions

File: src/writers.scala (hunk 3 above)
Comment (yannl35133) [reply src/writers.scala]: Simply using `symbol` worked, I guess `opaque symbol` will give better performance now

**Comment by yannl35133** [comment]
Command `isabelle dedukti_test` exports this "session" (I cannot quite understand the signification of the following yet, tbh)
```
session Dedukti_Import in "Ex/Import" = Pure +
  options [export_theory, export_proofs, record_proofs = 2]
  sessions HOL
  theories
    HOL.Inductive
```
and it then type-checks it with lambdapi.

Except the 17700 warnings about variables whose name can be dropped, type checking works well (and, apart from the issue on idents, always has)

**Comment by fblanqui** [comment]
Remark: you can remove warnings by using the option --no-warnings (or -w for short).

**Comment by yannl35133** [comment]
PR #6 is more advanced and better, closing this one
